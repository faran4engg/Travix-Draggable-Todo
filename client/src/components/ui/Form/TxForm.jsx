import React, { useState, Fragment } from "react";
import { Formik, Field } from "formik";
import { Form } from "semantic-ui-react";
import { connect } from "react-redux";
import { addCard, editCard } from "../../../redux/actions";
import PropTypes from "prop-types";
import * as yup from "yup";
import TxInput from "./Input/TxInput";
import TxTextArea from "./TextArea/TxTextArea";
import TxButton from "../Button/TxButton";

const userSchema = yup.object().shape({
  title: yup
    .string()
    .required()
    .min(5)
    .matches(/^[a-z\d\-_\s]+$/i),

  description: yup
    .string()
    .required()
    .min(5)
    .max(128)
});

const TxForm = props => {
  const { onCancel, dispatch, listName, editData } = props;
  const [formData, setFormData] = useState(editData);

  const handleSubmit = (values, actions) => {
    actions.setSubmitting(true);
    setFormData(values);

    editData
      ? dispatch(editCard(listName, { ...values, id: editData.id }))
      : dispatch(addCard(listName, values));

    actions.setSubmitting(false);
    onCancel();
  };

  return (
    <Fragment>
      <Formik
        initialValues={formData}
        onSubmit={handleSubmit}
        validationSchema={userSchema}
      >
        {formikProps =>
          !formikProps.isSubmitting ? (
            <Form onSubmit={formikProps.handleSubmit}>
              <Field
                label="Title"
                type="text"
                placeholder="Title"
                name="title"
                value={formikProps.values.title || ""}
                component={TxInput}
              />
              <Field
                label="Description"
                name="description"
                value={formikProps.values.description || ""}
                type="text"
                placeholder="Description"
                component={TxTextArea}
              />
              <TxButton
                type="submit"
                positive
                animated="vertical"
                disabled={!formikProps.isValid || !formikProps.dirty}
                content="Save"
                iconName="save"
              />
              <TxButton
                color="black"
                animated="vertical"
                content="Cancel"
                iconName="cancel"
                click={onCancel}
              />
            </Form>
          ) : null
        }
      </Formik>
    </Fragment>
  );
};

TxForm.propTypes = {
  onCancel: PropTypes.func,
  dispatch: PropTypes.func,
  listName: PropTypes.string,
  editData: PropTypes.object
};

export default connect()(TxForm);
