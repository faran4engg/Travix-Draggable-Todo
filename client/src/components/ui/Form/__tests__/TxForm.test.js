import React from "react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { shallow, mount } from "enzyme";
import TxForm from "../TxForm";

describe("TxForm", () => {
  it("should render correctly ", () => {
    const store = configureStore()();
    const component = shallow(
      <Provider store={store}>
        <TxForm
          editData={{
            id: "123",
            title: "Sample Title",
            description: "Sample Desc"
          }}
          listName={"todo"}
        />
      </Provider>
    );

    expect(component).toMatchSnapshot();
  });

  describe("when not passing editData prop", () => {
    it("should render empty form with blank fields ", () => {
      const store = configureStore()();
      const component = mount(
        <Provider store={store}>
          <TxForm />
        </Provider>
      );

      expect(component.find("TxInput").value).toEqual(undefined);
      expect(component.find("TxTextArea").value).toEqual(undefined);
    });
  });

  describe("when passing editData prop with data", () => {
    it("should render form with prepopulated values ", () => {
      const store = configureStore()();
      const component = mount(
        <Provider store={store}>
          <TxForm
            editData={{
              id: "123",
              title: "Sample",
              description: "Sample Desc"
            }}
          />
        </Provider>
      );

      expect(component.find("TxInput")).toHaveLength(1);
      expect(component.find("TxTextArea")).toHaveLength(1);
    });
  });
});
