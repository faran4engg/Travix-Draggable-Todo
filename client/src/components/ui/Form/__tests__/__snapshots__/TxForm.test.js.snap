// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TxForm should render correctly  1`] = `ShallowWrapper {}`;

exports[`TxForm when not passing editData prop should render empty form with blank fields  1`] = `ReactWrapper {}`;

exports[`TxForm when passing editData prop with data should render form with prepopulated values  1`] = `ShallowWrapper {}`;
